### PR TITLE
Feat/add tx query test

### PR DESCRIPTION
**Pull request by sotnikov-s**
https://p2pvalidator.atlassian.net/browse/LSC-126

running test:
```
yarn test:interchain_tx_query
```

please note that it's not necessary anymore to copy+paste contracts to this repo, they are kept updated in the contracts/artifacts/ folder

### CODE REVIEW

File: src/test_contracts/contracts/user_sample/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,48 @@
+[package]
+name = "user_sample"
+version = "0.1.0"
+authors = ["sotnikov-s <sergei.s@p2p.org>"]
```
Comment (sotnikov-s) [inline src/test_contracts/contracts/user_sample/Cargo.toml:4]: what do we want to put here as authors?

File: src/test_contracts/contracts/user_sample/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+[dependencies]
+cosmwasm-std = { version = "1.0.0", features = ["staking"] }
+schemars = "0.8.10"
+serde = { version = "1.0.103", default-features = false, features = ["derive"] }
+base64 = "0.13.0"
+
+# TODO: replace with sane imports once neutron-contracts repo sources are open
+interchain_queries = { path = "../../../../../neutron-contracts/packages/interchain_queries", default-features = false, version = "0.1.0"}
+neutron_sudo = { path = "../../../../../neutron-contracts/packages/neutron_sudo", default-features = false, version = "0.1.0"}
+neutron_bindings = {path = "../../../../../neutron-contracts/packages/bindings"}
```
Comment (pr0n00gler) [inline src/test_contracts/contracts/user_sample/Cargo.toml:43]: The repo is already open, so:
```suggestion
interchain_queries = { git = "https://github.com/neutron-org/neutron-contracts", default-features = false, version = "0.1.0"}
neutron_sudo = { git = "https://github.com/neutron-org/neutron-contracts", default-features = false, version = "0.1.0"}
neutron_bindings = { git = "https://github.com/neutron-org/neutron-contracts" }
```

And you don't need an additional mount in the `build_release.sh` script

**Review by pr0n00gler — CHANGES_REQUESTED** [review]

File: src/helpers/cosmos.ts [hunk 3]
```diff
@@ -81,12 +81,24 @@ export class CosmosWrapper {
...
-    const contractPath = process.env.CONTRACTS_PATH || './contracts/artifacts';
+  // storeContract stores the wasm code of the given contract on the blockchain.
+  async storeContract(fileName: string): Promise<CodeId> {
+    const contractsPath = process.env.CONTRACTS_PATH || './contracts/artifacts';
+    return this.storeWasm(contractsPath, fileName)
+  }
+
+  // storeTestContract stores the wasm code of the given test contract on the blockchain.
+  async storeTestContract(fileName: string): Promise<CodeId> {
+    const contractsPath = process.env.TEST_CONTRACTS_PATH || './src/test_contracts/artifacts';
```
Comment (pr0n00gler) [inline src/helpers/cosmos.ts:92]: What do you think of storing all the artifacts in the root `contracts/artifacts` folder?

File: src/test_contracts/contracts/user_sample/Cargo.toml (hunk 2 above)
Comment (sotnikov-s) [reply src/test_contracts/contracts/user_sample/Cargo.toml]: glad to know about it! done

**Review by pr0n00gler — COMMENTED** [review]
And since we already have `contracts` folder in the root, what do you think of moving all the test-contracts there?

So we can just do `make build` inside `contracts` and all the artifacts will be in one place

File: src/helpers/cosmos.ts (hunk 3 above)
Comment (foxpy) [reply src/helpers/cosmos.ts:92]: `contracts/artifacts` folder is a folder to manually copy artifacts from another repositories to. `src/test_contracts/artifacts` seems to be a folder for contracts which are only used in this repo and only for integration tests.

I personally don't see any problems here.

**Comment by foxpy** [comment]
@pr0n00gler I guess your threads here are not relevant anymore and can be closed.

File: src/helpers/cosmos.ts (hunk 3 above)
Comment (sotnikov-s) [reply src/helpers/cosmos.ts]: we've decided to get rid of the user sample contract in the integration tests repo and work with the neutron-contracts expanding the example contracts for integration tests

File: src/testcases/interchain_tx_query.test.ts [hunk 4]
```diff
@@ -0,0 +1,566 @@
...
+  });
+
+  const addr1: string = "neutron1fj6yqrkpw6fmp7f7jhj57dujfpwal4m25dafzx";
+  const addr2: string = "neutron14uxvu22lhrazyxadaqv5d6lswu0p276ll7hrkl";
+  const addr3: string = "neutron1m4n7gt76c4edkcpg698pwy56jkxqsre4gpgg26";
+  let addr1ExpectedBalance: number = 0;
+  let addr2ExpectedBalance: number = 0;
+  let addr3ExpectedBalance: number = 0;
+  const sendingToAddr1_1: number = 10000;
+  const sendingToAddr2_1: number = 5000;
```
Comment (ratik) [inline src/testcases/interchain_tx_query.test.ts:45]: maybe
```suggestion
  const amountToAddr1_1: number = 10000;
  const amountToAddr2_1: number = 5000;
```

File: src/testcases/interchain_tx_query.test.ts [hunk 5]
```diff
@@ -0,0 +1,566 @@
...
+    });
+    test("instantiate contract", async () => {
+      contractAddress = await cm.instantiate(codeId, "{}", "neutron_interchain_queries");
+      expect(contractAddress).toEqual(
+        "neutron14hj2tavq8fpesdwxxcu44rty3hh90vhujrvcmstl4zr3txmfvw9s5c2epq"
+      );
+    });
+  });
+
+  const addr1: string = "neutron1fj6yqrkpw6fmp7f7jhj57dujfpwal4m25dafzx";
```
Comment (ratik) [inline src/testcases/interchain_tx_query.test.ts:38]: do you mind moving this block of vars/consts to the top of describe block? (and similar following ones)

File: src/helpers/cosmos.ts [hunk 6]
```diff
@@ -1,5 +1,6 @@
 import { promises as fsPromise } from 'fs';
 import { cosmosclient, rest, proto } from '@cosmos-client/core';
+import { AccAddress } from '@cosmos-client/core/cjs/types';
```
Comment (ratik) [inline src/helpers/cosmos.ts:3]: seems unused

File: src/testcases/interchain_tx_query.test.ts (hunk 5 above)
Comment (sotnikov-s) [reply src/testcases/interchain_tx_query.test.ts:38]: they all are used inside other describe scopes below, so no, I can not :)

**Comment by foxpy** [comment]
Wait, I have missed the part when we have started storing wasm artifacts in this repo. What is the reasoning to do so?

**Comment by sotnikov-s** [comment]
> Wait, I have missed the part when we have started storing wasm artifacts in this repo. What is the reasoning to do so?

why not? to make the repo ready to use and keep the up to date contracts here, otherwise each of us will have to compile contracts each time he wants to run tests in this or that branch

**Comment by foxpy** [comment]
> otherwise each of us will have to compile contracts

they are already compiled, you only have to `cp` them.

also, we didn't ship compiled contracts in our previous integration tests framework anyway.

**Comment by sotnikov-s** [comment]
> > otherwise each of us will have to compile contracts
> 
> they are already compiled, you only have to `cp` them.
> 
> also, we didn't ship compiled contracts in our previous integration tests framework anyway.

yeah, sorry, I meant `copy`. you will have to switch to a possible branch that enhances the contract logic, copy the wasm file to the integration tests repo, etc. I frankly see no meaning in not having up to date contracts here and no harm in having them. you anyway store the latest contract you used in this repo, so why not push it so others will not have to do the copying if any occurred?

**Comment by foxpy** [comment]
> so why not push it so others will not have to do the copying if any occurred?

Then I at least ask you to store `checksums.txt` as well. It would simplify things further.

**Review by foxpy — CHANGES_REQUESTED** [review]
Tests are oudated, transactions filter logic has been changed, now you have to fetch `main` in contracts and neutron repos and patch our tests.

**Review by pr0n00gler — APPROVED** [review]

File: src/testcases/interchain_tx_query.test.ts [hunk 7]
```diff
@@ -0,0 +1,622 @@
...
+          recipient: watchedAddr1,
+          sender: cm2.wallet.address.toString(),
+          denom: cm2.denom,
+          amount: addr1ExpectedBalance.toString(),
+        },
+      ]);
+    });
+
+    test('handle failed transfer', async () => {
+      const res = await cm2.msgSend(watchedAddr1, '99999999999999'); // the amount is greater than the sender has
```
Comment (foxpy) [inline src/testcases/interchain_tx_query.test.ts:141]:
```suggestion
      const res = await cm2.msgSend(watchedAddr1, '99999999999999'); // the amount is greater than the sender has
      expect(res.code).not.toEqual(0);
```

File: src/testcases/interchain_tx_query.test.ts [hunk 8]
```diff
@@ -0,0 +1,622 @@
...
+      let balances = await cm2.queryBalances(watchedAddr1);
+      expect(balances.balances).toEqual([]);
+      const res = await cm2.msgSend(watchedAddr1, amountToAddr1_1.toString());
+      expect(res.code).toEqual(0);
+      balances = await cm2.queryBalances(watchedAddr1);
+      expect(balances.balances).toEqual([
+        { amount: addr1ExpectedBalance.toString(), denom: cm2.denom },
+      ]);
+
+      await wait(query1UpdatePeriod * BLOCK_TIME);
```
Comment (foxpy) [inline src/testcases/interchain_tx_query.test.ts:92]:
```suggestion
      await wait((query1UpdatePeriod + 1) * BLOCK_TIME);
```

File: src/testcases/interchain_tx_query.test.ts [hunk 9]
```diff
@@ -0,0 +1,622 @@
...
+      addr2ExpectedBalance += amountToAddr2_1;
+      let balances = await cm2.queryBalances(differentAddr);
+      expect(balances.balances).toEqual([]);
+      const res = await cm2.msgSend(differentAddr, amountToAddr2_1.toString());
+      expect(res.code).toEqual(0);
+      balances = await cm2.queryBalances(differentAddr);
+      expect(balances.balances).toEqual([
+        { amount: addr2ExpectedBalance.toString(), denom: cm2.denom },
+      ]);
+      await wait(query1UpdatePeriod * BLOCK_TIME);
```
Comment (foxpy) [inline src/testcases/interchain_tx_query.test.ts:119]:
```suggestion
      await wait((query1UpdatePeriod + 1) * BLOCK_TIME);
```

File: src/testcases/interchain_tx_query.test.ts [hunk 10]
```diff
@@ -0,0 +1,622 @@
...
+    });
+
+    test('handle failed transfer', async () => {
+      const res = await cm2.msgSend(watchedAddr1, '99999999999999'); // the amount is greater than the sender has
+      expect(res.txhash?.length).toBeGreaterThan(0); // hash is not empty thus tx executed
+      const balances = await cm2.queryBalances(watchedAddr1);
+      expect(balances.balances).toEqual([
+        { amount: addr1ExpectedBalance.toString(), denom: cm2.denom }, // balance hasn't changed thus tx failed
+      ]);
+      await wait(query1UpdatePeriod * BLOCK_TIME);
```
Comment (foxpy) [inline src/testcases/interchain_tx_query.test.ts:147]:
```suggestion
      await wait((query1UpdatePeriod + 1) * BLOCK_TIME);
```

File: src/testcases/interchain_tx_query.test.ts [hunk 11]
```diff
@@ -0,0 +1,622 @@
...
+      expect(query.registered_query.update_period).toEqual(query2UpdatePeriod);
+    });
+
+    test('handle callback on a past sending', async () => {
+      const balances = await cm2.queryBalances(watchedAddr2);
+      expect(balances.balances).toEqual([
+        { amount: addr2ExpectedBalance.toString(), denom: cm2.denom },
+      ]);
+
+      await wait(query2UpdatePeriod * BLOCK_TIME);
```
Comment (foxpy) [inline src/testcases/interchain_tx_query.test.ts:202]:
```suggestion
      await wait((query2UpdatePeriod + 1) * BLOCK_TIME);
```

File: src/testcases/interchain_tx_query.test.ts [hunk 12]
```diff
@@ -0,0 +1,622 @@
...
+      expect(res.code).toEqual(0);
+      balances = await cm2.queryBalances(watchedAddr3);
+      expect(balances.balances).toEqual([
+        { amount: addr3ExpectedBalance.toString(), denom: cm2.denom },
+      ]);
+      let deposits = await queryRecipientTxs(cm, contractAddress, watchedAddr3);
+      // update time hasn't come yet despite the fact the sent funds are already on the account
+      expect(deposits.transfers).toEqual([]);
+
+      await wait(query3UpdatePeriod * BLOCK_TIME);
```
Comment (foxpy) [inline src/testcases/interchain_tx_query.test.ts:265]:
```suggestion
      await wait((query3UpdatePeriod + 1) * BLOCK_TIME);
```

File: src/testcases/interchain_tx_query.test.ts [hunk 13]
```diff
@@ -0,0 +1,622 @@
...
+      expect(deposits.transfers).toEqual([
+        {
+          recipient: watchedAddr3,
+          sender: cm2.wallet.address.toString(),
+          denom: cm2.denom,
+          amount: (addr3ExpectedBalance - amountToAddr3_2).toString(),
+        },
+      ]);
+
+      await wait(query3UpdatePeriod * BLOCK_TIME);
```
Comment (foxpy) [inline src/testcases/interchain_tx_query.test.ts:296]:
```suggestion
      await wait((query3UpdatePeriod + 1) * BLOCK_TIME);
```

File: src/testcases/interchain_tx_query.test.ts [hunk 14]
```diff
@@ -0,0 +1,622 @@
...
+      ]);
+      balances = await cm2.queryBalances(watchedAddr3);
+      expect(balances.balances).toEqual([
+        { amount: addr3ExpectedBalance.toString(), denom: cm2.denom },
+      ]);
+    });
+
+    test('check transfers handled', async () => {
+      await wait(
+        Math.max(query1UpdatePeriod, query2UpdatePeriod, query3UpdatePeriod) *
```
Comment (foxpy) [inline src/testcases/interchain_tx_query.test.ts:354]:
```suggestion
        (Math.max(query1UpdatePeriod, query2UpdatePeriod, query3UpdatePeriod) +
          1) *
```

File: src/testcases/interchain_tx_query.test.ts [hunk 15]
```diff
@@ -0,0 +1,622 @@
...
+      let balances = await cm2.queryBalances(watchedAddr4);
+      expect(balances.balances).toEqual([]);
+      const res = await cm2.msgSend(watchedAddr4, amountToAddr4_1.toString());
+      expect(res.code).toEqual(0);
+      balances = await cm2.queryBalances(watchedAddr4);
+      expect(balances.balances).toEqual([
+        { amount: addr4ExpectedBalance.toString(), denom: cm2.denom },
+      ]);
+
+      await wait(query4UpdatePeriod * BLOCK_TIME);
```
Comment (foxpy) [inline src/testcases/interchain_tx_query.test.ts:500]:
```suggestion
      await wait((query4UpdatePeriod + 1) * BLOCK_TIME);
```

File: src/testcases/interchain_tx_query.test.ts [hunk 16]
```diff
@@ -0,0 +1,622 @@
...
+          recipient: watchedAddr4,
+          sender: cm2.wallet.address.toString(),
+          denom: cm2.denom,
+          amount: addr4ExpectedBalance.toString(),
+        },
+      ]);
+      deposits = await queryRecipientTxs(cm, contractAddress, watchedAddr5);
+      expect(deposits.transfers).toEqual([]);
+
+      await wait(query5UpdatePeriod * BLOCK_TIME);
```
Comment (foxpy) [inline src/testcases/interchain_tx_query.test.ts:514]:
```suggestion
      await wait((query5UpdatePeriod + 1) * BLOCK_TIME);
```

**Review by foxpy — CHANGES_REQUESTED** [review]
Please apply suggested changes to fix race conditions.

**Review by foxpy — DISMISSED** [review]
Apparently, the aforementioned race conditions are only reproducible on my machine (?). As much as I would love to see them fixed (it is enough just to apply my suggestions), I am approving this PR to prevent further review delays.

**Comment by sotnikov-s** [comment]
> Apparently, the aforementioned race conditions are only reproducible on my machine (?). As much as I would love to see them fixed (it is enough just to apply my suggestions), I am approving this PR to prevent further review delays.

I've run the tests 25+ times and haven't encountered any race condition. Race conditions ceased to be my problem since stateful relayer came into play. But anyway I appreciate your request and I've increased the pause duration as you suggested. Please take a look now

**Comment by foxpy** [comment]
After extensive testing I have discovered these fixes don't work :|

They just make fails more rare.

**Review by foxpy — APPROVED** [review]
Anyway, approving this as it is

**Review by foxpy — APPROVED** [review]

**Pull request merged**